text/0000-build-std.md

```diff
@@ -0,0 +1,4582 @@
+vendored, Cargo will "lazily" download them the first time `build-std` is used
+but this may happen multiple times if they are cleaned from its cache without
+upgrading the toolchain version.
+
+See
+[*Why use the lockfile of the `rust-src` component?*][why-use-the-lockfile-of-the-rust-src-component]
+
+↩ [*Vendored `rust-src`*][vendored-rust-src]
+
+### Why not check if `rust-src` has been modified?
+[rationale-src-modifications]: #why-not-check-if-rust-src-has-been-modified
+
+It is likely that any protections implemented to check that the sources in
+`rust-src` have not been modified could be trivially bypassed.
+
+Any crate that depends on `rust-src` having been modified would not be usable
+when published to crates.io as the required modifications will obviously not be
+included.
+
+↩ [*Vendored `rust-src`*][vendored-rust-src]
+
+### Why allow building from the sysroot with implied `RUSTC_BOOTSTRAP`?
+[rationale-implied-bootstrap]: #why-allow-building-from-the-sysroot-with-implied-rustc_bootstrap
+
+Cargo needs to be able to build the standard library crates, which inherently
+require a nightly toolchain. It could set `RUSTC_BOOTSTRAP` internally to do
+this with a stable toolchain, however this is a shared requirement with other
+build systems that wish to build an unmodified standard library and want to work
+on stable toolchains.
+
+For example, Rust's project goal to enable Rust for Linux to build using only a
+stable toolchain would require that it be possible to build `core` without
+nightly.
+
+It is not sufficient for rustc to special-case the `core`, `alloc` and `std`
+crate names as when being built as part of the standard library, dependencies of
+the standard library also use unstable features and so these crate would also
+need such special-casing, which is not practical.
+
+↩ [*Building the standard library on a stable toolchain*][building-the-standard-library-on-a-stable-toolchain]
+
+### Why not use weak linkage for `compiler-builtins/mem` symbols?
+[rationale-no-weak-linkage]: #why-not-use-weak-linkage-for-compiler-builtinsmem-symbols
+
+Since [compiler-builtins#411], the relevant symbols in `compiler_builtins`
+already have weak linkage. However, it is nevertheless not possible to simply
+remove the `mem` feature and have the symbols always be present.
+
+Some targets, such as those based on MinGW, do not have sufficient support for
+weak definitions (at least with the default linker). Furthermore, weak linkage
+has precedence over shared libraries and the symbols of a dynamically-linked
+`libc` should be preferred over `compiler_builtins`'s symbols.
+
+↩ [*`compiler-builtins-mem`*][compiler-builtins-mem]
+
+### Why permit breakage of nightly build-std users using tier three targets?
+[rationale-breakage]: #why-permit-breakage-of-nightly-build-std-users-using-tier-three-targets
+
+As this use case would previously have been using the unstable
+`-Zbuild-std=core`, this user must be on a nightly toolchain. This proposal
```

Comment:
I don't entirely agree with this. I believe there are people building `rustc` with their own targets, and shipping it themselves, thus having a "stable" version of these targets.

Reply:
I don't think these targets are "stable" in the sense that the Rust project provides any guarantees for them. If a user is maintaining a downstream fork then they're taking on responsibility to keep that working. I don't think it's especially reasonable to constrain what we can do to save those users from having to fix their fork.